### DISCUSSION TITLE
markdown inside of a shape

### DISCUSSION THREAD

**Discussion by ackstorm23** [terrastruct/d2#984] [Q&A]
It would seem that you cannot use markdown formatting while also using a shape?

this will render markdown, but the default shape is then disabled

```
a: |md
# title
- bullet
|
```

if you try to define a shape, that shape is then used, but the markdown is lost

```
a: |md
# title
- bullet
|
a.shape: circle
```

Is this on purpose?  Is this  bug?

**Comment by alixander** [comment]
@ackstorm23 it's not a bug but it also will be doable soon. currently the behavior is that markdown labels have an implicit `shape: text`, but there's no UX reason why this what you're doing shouldn't be possible from a user perspective, so we'll support it before v1.

for now you can add a child markdown:

https://play.d2lang.com/?script=HM6xCYAwFEXR_k1xibUL_G00pgh8G3mCYNxdSHuqswWl8ApmPKj9qtkEfhyM8xDAgrsnw8p-ZzYLhj79AQAA__8%3D&layout=dagre&

**Reply by muteboy** [reply]
That link doesn't work - can you give another, or paste the example here? Thank you

**Reply by alixander** [reply]
https://play.d2lang.com/?script=SrRSUFKq5lJQSM7PK0nNK7FSqMlN4VJQUFBQVgjJLMlJBbOT8lMquRQUarhquQABAAD__w%3D%3D&

**Comment by gwhitney** [comment]
Status on this? There doesn't currently seem to be any way to get a shape label to have multiple styles (some bold, some not). I tried adding a child markdown as suggested here, but then the shape is a container and there's so much padding that it makes the text in the markdown child too small to read, and then if I increase the font size in the markdown, the shape just becomes huge with all the padding and dwarfs all my other nodes. And then if I increase the size of the other nodes, the whole diagram just rescales and again the markdown child is unreadable, so it becomes a vicious cycle. Or is there any way to dial down the padding on a container? That would be a really helpful style setting...

**Reply by alixander** [reply]
I [played around](https://github.com/terrastruct/d2/pull/2204/files) with allowing markdown as labels, but it looks very similar to containers-with-markdown-children, just with zero padding.  

<img width="178" alt="Screenshot 2024-11-08 at 3 04 46 PM" src="https://github.com/user-attachments/assets/8556b96e-e970-4b11-986f-45a017f69427">

So I figure I should add some padding, which then turns into the same as the current solution.

So this issue should just be to control padding more: https://github.com/terrastruct/d2/issues/1740

I'll find some time to work on this this sprint.

**Reply by alixander** [reply]
@gwhitney Can you actually provide an example of where this is obviously a problem? I've seen markdown inside shapes being used like this and they seem okay tome.

https://play.d2lang.com/?script=jFTRbqs4EH3nK0b0LRJsoMmqy1vVbLeVmmwUdlupulLr4EmwCjaynRBum3-_wpiGtCg3DzzgmTPnHHtmtkSqCN4dABp6ieArtm5-ATJSiY32kK8ZxwgwezPHOsUcPUYjCIdDB2Dv7B3KJCaaCR4BFSV3wEkyohTa0oNDySVmEbiu-VMpKTCCGkr4OsPmUFcZtvkASyEpSk8SyjYqgmBozvdO823zNvMLTGkp3tCjRKURjDqYvQMO4xolR92kWxVJJjbUOUi8t0mugUix0SgbAO6KjHDS2P3Iqal-cQELk2P_LuBObBRGEPwV-sGfV37gB5-hyfS5EwgvbciDfwvkT1JDeOkPx_7Ynk4msxi0gEKKLaOWwoNbIUsiKeNruBqNLuuMG0Jp5QB8GNHbjPCXtJHx3rFmlMHjw_UMvI6-4R_hyD0Aaf7zGDaZPn8D1UKPYYbPL9mKnb6tJ3bL4DpJUCmYC8b15-0cFIXjkfUax_eTCAaDuUStK1hlFayEBAI10WDQY9nPq5ecJEsh3k4LeSCFFsXhbe5u5pZ02sBhLkUfQaFGpyvPM1LF2pzbHuwQfBzdtM83yelis_9vwIO7qkC5ZUrInuuyw-HVgne52MHj371EflK3SUNn5jSCbX6C2qSDBwvcolRoylc9_OOWf4JLRjg8TqFkOoWFEDpr3pnmhLciN8uMqbRu30JIbXrYhp5YRhMiKfz3EJt3fh34uCN5kaGfiPy135VU6nxPUql6hCSqdBHHfWaCc830SCnL8nwpZVnWnnGpmMY-JWG_klTrgn7yt0sNPA_abeXex3OYkRx_8AD-WRbKtYusTupuB9fMdfAtbHaADQ6H7tcJPy5zPHHtQj7exPv9b2o0Q3UafNTGXa1tG5yTVpZlj9vueLjzui3tZq2703V-BQAA__8%3D&

![d2 (2)](https://github.com/user-attachments/assets/dd035e55-ae73-4de6-89ae-dd5b81790aa1)

**Reply by gwhitney** [reply]
Well, I've moved on from this because of this issue, but here's more-or-less the diagram I had:

https://play.d2lang.com/?script=nJJPi9swEMXv-hSDc0mgYdurDr2kl94COZZCJtKsPa0imdE46Tbxdy-S1ssSelqDsf78nmbeszwLOeUULQj3gxoXMGfKFm4G4Ew6JN_GAFlfAi0TgFMST7IV9DxlC18-1_XZlHc25qAoaqHrqmDRPnMIFk5hogLmAUey4FhcIANwZa9DPWo23yO_kzdwRMEQKKRe8GwAHI6t9_vZ1-qr1Qp2QqgEF84TBv5LAun0i5xW4Hg86p82HCWNJPoCHFm5oflT3XIpZpXJaZJFZQDub7Zg-xVaf0vhEXuCkNA_GXhf2g0YezJ3c9jvPuRmj4Jn0noUud_50UV7OHq-sJ8wvC5eMLBHpfUGnqdY_3B-tFIcFCeH_c4cSKfRQpfLd73p4FYvgn29AnPpv7Jl33wTvFrovOD1v2yBCl24Ci8TCzdF6Um3KJKuA6G30P3Imsb15smlqBwnWm9-dvObriXXgqyZ1Tg68y8AAP__&

I just couldn't find any way to get all of the labels & text readable, even though it seems like there should be plenty of space. Increasing the font-size on the edge labels would in effect shrink the parallelogram labels, I think because of a phenomenon of making the edges longer and then the whole thing rescaling into the space available, and so then I would jack up the font size in the markdown children, and then the setup() and draw() labels would be tiny, so I would have to try to increase their font sizes, and then the edge labels would look tiny again. So it just became a vicious cycle. There was a very frustrating lack of ability to Just. Make. Text. Bigger! (leaving the size/layout of everything else alone). Best of luck.

**Reply by alixander** [reply]
I understand, thanks for coming back to leave the diagram!
